### ISSUE TITLE
DL4J Keras Import: tf.keras support

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#8348]
TensorFlow 2 has 'build in' Keras model saving support:
https://www.tensorflow.org/tutorials/keras/save_and_load#hdf5_format

It looks like there are some format differences between this and the standard Keras model saving.

tf.keras is not yet tested with DL4J Keras import, but we know that there are some issues as of 1.0.0-beta5. For example:
https://github.com/eclipse/deeplearning4j/issues/8338
https://github.com/eclipse/deeplearning4j/issues/8344

Until this has been resolved, using the "standard" (non-TF) Keras - which is well tested - is one workaround.
Also using SameDiff's TensorFlow import support is another workaround (from TF frozen models).

**Comment by victoriest** [comment]
Detail of #8338 
```
Exception in thread "main" org.deeplearning4j.exception.DL4JInvalidConfigException: Invalid configuration for layer (idx=-1, name=max_pooling2d_2, type=SubsamplingLayer) for height dimension:  Invalid input configuration for kernel height. Require 0 < kH <= inHeight + 2*padH; got (kH=16, inHeight=14, padH=0)
Input type = InputTypeConvolutional(h=14,w=14,c=192), kernel = [16, 16], strides = [16, 16], padding = [0, 0], layer size (output channels) = 192, convolution mode = Same
	at org.deeplearning4j.nn.conf.layers.InputTypeUtil.getOutputTypeCnnLayers(InputTypeUtil.java:327)
	at org.deeplearning4j.nn.conf.layers.SubsamplingLayer.getOutputType(SubsamplingLayer.java:153)
	at org.deeplearning4j.nn.modelimport.keras.layers.pooling.KerasPooling2D.getOutputType(KerasPooling2D.java:95)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.inferOutputTypes(KerasModel.java:304)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.<init>(KerasModel.java:179)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.<init>(KerasModel.java:96)
	at org.deeplearning4j.nn.modelimport.keras.utils.KerasModelBuilder.buildModel(KerasModelBuilder.java:307)
	at org.deeplearning4j.nn.modelimport.keras.KerasModelImport.importKerasModelAndWeights(KerasModelImport.java:172)
......
```

Code of model define:

```python
    def __gen__rcnn_model():
        """
        RCNN, ref: https://github.com/JimLee4530/RCNN
        :return:
        """
        input_img = Input(shape=(INPUT_IMG_HEIGHT, INPUT_IMG_WIDTH, 3), name="input")
        conv1 = Conv2D(filters=192, kernel_size=[5, 5], strides=(1, 1), padding='same', activation='relu')(input_img)

        rconv1 = OcrModel.__rcl_block(192, conv1)
        dropout1 = Dropout(0.2)(rconv1)
        rconv2 = OcrModel.__rcl_block(192, dropout1)
        maxpooling_1 = MaxPooling2D((2, 2), strides=(2, 2), padding='same')(rconv2)
        dropout2 = Dropout(0.2)(maxpooling_1)
        rconv3 = OcrModel.__rcl_block(192, dropout2)
        dropout3 = Dropout(0.2)(rconv3)
        rconv4 = OcrModel.__rcl_block(192, dropout3)

        out = MaxPool2D((16, 16), strides=(16, 16), padding='same')(rconv4)
        flatten = Flatten()(out)
        prediction = Dense(11, activation='softmax')(flatten)

        model = Model(inputs=input_img, outputs=prediction)
        adam = optimizers.Adam(lr=0.0005, beta_1=0.9, beta_2=0.999, epsilon=1e-08)
        model.compile(optimizer=adam, loss='categorical_crossentropy', metrics=['accuracy'])
```

Summary by ```model.summary()```

```
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
input (InputLayer)              (None, 28, 28, 3)    0                                            
__________________________________________________________________________________________________
conv2d_1 (Conv2D)               (None, 28, 28, 192)  14592       input[0][0]                      
__________________________________________________________________________________________________
conv2d_2 (Conv2D)               (None, 28, 28, 192)  331968      conv2d_1[0][0]                   
__________________________________________________________________________________________________
batch_normalization_1 (BatchNor (None, 28, 28, 192)  768         conv2d_2[0][0]                   
__________________________________________________________________________________________________
conv2d_3 (Conv2D)               (None, 28, 28, 192)  331968      batch_normalization_1[0][0]      
__________________________________________________________________________________________________
add_1 (Add)                     (None, 28, 28, 192)  0           conv2d_2[0][0]                   
                                                                 conv2d_3[0][0]                   
__________________________________________________________________________________________________
batch_normalization_2 (BatchNor (None, 28, 28, 192)  768         add_1[0][0]                      
__________________________________________________________________________________________________
conv2d_4 (Conv2D)               (None, 28, 28, 192)  331968      batch_normalization_2[0][0]      
__________________________________________________________________________________________________
add_2 (Add)                     (None, 28, 28, 192)  0           conv2d_2[0][0]                   
                                                                 conv2d_4[0][0]                   
__________________________________________________________________________________________________
batch_normalization_3 (BatchNor (None, 28, 28, 192)  768         add_2[0][0]                      
__________________________________________________________________________________________________
conv2d_5 (Conv2D)               (None, 28, 28, 192)  331968      batch_normalization_3[0][0]      
__________________________________________________________________________________________________
add_3 (Add)                     (None, 28, 28, 192)  0           conv2d_2[0][0]                   
                                                                 conv2d_5[0][0]                   
__________________________________________________________________________________________________
batch_normalization_4 (BatchNor (None, 28, 28, 192)  768         add_3[0][0]                      
__________________________________________________________________________________________________
dropout_1 (Dropout)             (None, 28, 28, 192)  0           batch_normalization_4[0][0]      
__________________________________________________________________________________________________
conv2d_6 (Conv2D)               (None, 28, 28, 192)  331968      dropout_1[0][0]                  
__________________________________________________________________________________________________
batch_normalization_5 (BatchNor (None, 28, 28, 192)  768         conv2d_6[0][0]                   
__________________________________________________________________________________________________
conv2d_7 (Conv2D)               (None, 28, 28, 192)  331968      batch_normalization_5[0][0]      
__________________________________________________________________________________________________
add_4 (Add)                     (None, 28, 28, 192)  0           conv2d_6[0][0]                   
                                                                 conv2d_7[0][0]                   
__________________________________________________________________________________________________
batch_normalization_6 (BatchNor (None, 28, 28, 192)  768         add_4[0][0]                      
__________________________________________________________________________________________________
conv2d_8 (Conv2D)               (None, 28, 28, 192)  331968      batch_normalization_6[0][0]      
__________________________________________________________________________________________________
add_5 (Add)                     (None, 28, 28, 192)  0           conv2d_6[0][0]                   
                                                                 conv2d_8[0][0]                   
__________________________________________________________________________________________________
batch_normalization_7 (BatchNor (None, 28, 28, 192)  768         add_5[0][0]                      
__________________________________________________________________________________________________
conv2d_9 (Conv2D)               (None, 28, 28, 192)  331968      batch_normalization_7[0][0]      
__________________________________________________________________________________________________
add_6 (Add)                     (None, 28, 28, 192)  0           conv2d_6[0][0]                   
                                                                 conv2d_9[0][0]                   
__________________________________________________________________________________________________
batch_normalization_8 (BatchNor (None, 28, 28, 192)  768         add_6[0][0]                      
__________________________________________________________________________________________________
max_pooling2d_1 (MaxPooling2D)  (None, 14, 14, 192)  0           batch_normalization_8[0][0]      
__________________________________________________________________________________________________
dropout_2 (Dropout)             (None, 14, 14, 192)  0           max_pooling2d_1[0][0]            
__________________________________________________________________________________________________
conv2d_10 (Conv2D)              (None, 14, 14, 192)  331968      dropout_2[0][0]                  
__________________________________________________________________________________________________
batch_normalization_9 (BatchNor (None, 14, 14, 192)  768         conv2d_10[0][0]                  
__________________________________________________________________________________________________
conv2d_11 (Conv2D)              (None, 14, 14, 192)  331968      batch_normalization_9[0][0]      
__________________________________________________________________________________________________
add_7 (Add)                     (None, 14, 14, 192)  0           conv2d_10[0][0]                  
                                                                 conv2d_11[0][0]                  
__________________________________________________________________________________________________
batch_normalization_10 (BatchNo (None, 14, 14, 192)  768         add_7[0][0]                      
__________________________________________________________________________________________________
conv2d_12 (Conv2D)              (None, 14, 14, 192)  331968      batch_normalization_10[0][0]     
__________________________________________________________________________________________________
add_8 (Add)                     (None, 14, 14, 192)  0           conv2d_10[0][0]                  
                                                                 conv2d_12[0][0]                  
__________________________________________________________________________________________________
batch_normalization_11 (BatchNo (None, 14, 14, 192)  768         add_8[0][0]                      
__________________________________________________________________________________________________
conv2d_13 (Conv2D)              (None, 14, 14, 192)  331968      batch_normalization_11[0][0]     
__________________________________________________________________________________________________
add_9 (Add)                     (None, 14, 14, 192)  0           conv2d_10[0][0]                  
                                                                 conv2d_13[0][0]                  
__________________________________________________________________________________________________
batch_normalization_12 (BatchNo (None, 14, 14, 192)  768         add_9[0][0]                      
__________________________________________________________________________________________________
dropout_3 (Dropout)             (None, 14, 14, 192)  0           batch_normalization_12[0][0]     
__________________________________________________________________________________________________
conv2d_14 (Conv2D)              (None, 14, 14, 192)  331968      dropout_3[0][0]                  
__________________________________________________________________________________________________
batch_normalization_13 (BatchNo (None, 14, 14, 192)  768         conv2d_14[0][0]                  
__________________________________________________________________________________________________
conv2d_15 (Conv2D)              (None, 14, 14, 192)  331968      batch_normalization_13[0][0]     
__________________________________________________________________________________________________
add_10 (Add)                    (None, 14, 14, 192)  0           conv2d_14[0][0]                  
                                                                 conv2d_15[0][0]                  
__________________________________________________________________________________________________
batch_normalization_14 (BatchNo (None, 14, 14, 192)  768         add_10[0][0]                     
__________________________________________________________________________________________________
conv2d_16 (Conv2D)              (None, 14, 14, 192)  331968      batch_normalization_14[0][0]     
__________________________________________________________________________________________________
add_11 (Add)                    (None, 14, 14, 192)  0           conv2d_14[0][0]                  
                                                                 conv2d_16[0][0]                  
__________________________________________________________________________________________________
batch_normalization_15 (BatchNo (None, 14, 14, 192)  768         add_11[0][0]                     
__________________________________________________________________________________________________
conv2d_17 (Conv2D)              (None, 14, 14, 192)  331968      batch_normalization_15[0][0]     
__________________________________________________________________________________________________
add_12 (Add)                    (None, 14, 14, 192)  0           conv2d_14[0][0]                  
                                                                 conv2d_17[0][0]                  
__________________________________________________________________________________________________
batch_normalization_16 (BatchNo (None, 14, 14, 192)  768         add_12[0][0]                     
__________________________________________________________________________________________________
max_pooling2d_2 (MaxPooling2D)  (None, 1, 1, 192)    0           batch_normalization_16[0][0]     
__________________________________________________________________________________________________
flatten_1 (Flatten)             (None, 192)          0           max_pooling2d_2[0][0]            
__________________________________________________________________________________________________
dense_1 (Dense)                 (None, 11)           2123        flatten_1[0][0]                  
==================================================================================================
Total params: 5,340,491
Trainable params: 5,334,347
Non-trainable params: 6,144
__________________________________________________________________________________________________
```

**Comment by AlexDBlack** [comment]
@victoriest ok, so this is pretty much what I expected, and mentioned in the other issue: keras must be adding some implicit padding here.
We (correctly, IMO) consider a kernel size larger than the input size to be invalid. You have a 16x16 kernel operating on 14x14 input, which doesn't really make sense.
That said, for import compatibility, we may need to allow Keras' way of implicitly adding padding.

`out = MaxPool2D((16, 16), strides=(16, 16), padding='same')(rconv4)`

```
batch_normalization_16 (BatchNo (None, 14, 14, 192)  768         add_12[0][0]                     
max_pooling2d_2 (MaxPooling2D)  (None, 1, 1, 192)    0           batch_normalization_16[0][0]
```

In the mean time, here's 2 ways that are more correct architectures:
- Use a 14x14 kernel on your 14x14 input
- Better: Use a GlobalMaxPooling2D layer instead of your MaxPool2D + flatten https://keras.io/layers/pooling/

Either of those should import fine.

I'll open a separate issue for this Keras edge case, it's unrelated to tf.keras support for DL4J Keras import

**Comment by victoriest** [comment]
The solution is worked(14*14 kernel). im so appreciated and thanks for the reply.

**Comment by ptamas88** [comment]
Hello @treo @AlexDBlack ,
is there any ETA for tf.keras support?
We are developing models in tf2 and would like to handle serving in java with dl4j natively.
Thanks

**Comment by AlexDBlack** [comment]
@ptamas88 Thanks for the reminder. It's actually available on snapshots now:, and will be in the next release https://github.com/KonduitAI/deeplearning4j/pull/258
https://deeplearning4j.konduit.ai/config/config-snapshots
Some additional work is currently underway to improve test coverage though.

**Comment by ptamas88** [comment]
@AlexDBlack Thanks for the response!

**Comment by rhorvath37** [comment]
Hello @AlexDBlack ,

The attached file was tested with snapshot but we got the same exception:

Exception in thread "main" java.lang.RuntimeException: org.nd4j.shade.protobuf.InvalidProtocolBufferException$InvalidWireTypeException: Protocol message tag had invalid wire type.
	at org.nd4j.imports.graphmapper.tf.TFGraphMapper.importGraph(TFGraphMapper.java:137)
	at org.nd4j.imports.graphmapper.tf.TFGraphMapper.importGraph(TFGraphMapper.java:86)
	at org.nd4j.imports.graphmapper.tf.TFGraphMapper.importGraph(TFGraphMapper.java:72)
	at org.nd4j.autodiff.samediff.SameDiff.importFrozenTF(SameDiff.java:5714)
	at vvk.dl4j.test.LoadTensorFlow.getGraph(LoadTensorFlow.java:36)
	at vvk.dl4j.test.LoadTensorFlow.main(LoadTensorFlow.java:30)
Caused by: org.nd4j.shade.protobuf.InvalidProtocolBufferException$InvalidWireTypeException: Protocol message tag had invalid wire type.
	at org.nd4j.shade.protobuf.InvalidProtocolBufferException.invalidWireType(InvalidProtocolBufferException.java:111)
	at org.nd4j.shade.protobuf.UnknownFieldSet$Builder.mergeFieldFrom(UnknownFieldSet.java:557)
	at org.nd4j.shade.protobuf.GeneratedMessageV3.parseUnknownField(GeneratedMessageV3.java:298)
	at org.tensorflow.framework.GraphDef.<init>(GraphDef.java:98)
	at org.tensorflow.framework.GraphDef.<init>(GraphDef.java:13)
	at org.tensorflow.framework.GraphDef$1.parsePartialFrom(GraphDef.java:1569)
	at org.tensorflow.framework.GraphDef$1.parsePartialFrom(GraphDef.java:1563)
	at org.nd4j.shade.protobuf.AbstractParser.parsePartialFrom(AbstractParser.java:215)
	at org.nd4j.shade.protobuf.AbstractParser.parseFrom(AbstractParser.java:232)
	at org.nd4j.shade.protobuf.AbstractParser.parseFrom(AbstractParser.java:237)
	at org.nd4j.shade.protobuf.AbstractParser.parseFrom(AbstractParser.java:48)
	at org.nd4j.shade.protobuf.GeneratedMessageV3.parseWithIOException(GeneratedMessageV3.java:317)
	at org.tensorflow.framework.GraphDef.parseFrom(GraphDef.java:462)
	at org.nd4j.imports.graphmapper.tf.TFGraphMapper.importGraph(TFGraphMapper.java:135)
	... 5 more

Could you please help what is the issue?
[frozen_graph.zip](https://github.com/eclipse/deeplearning4j/files/4540540/frozen_graph.zip)

Thak You

**Comment by AlexDBlack** [comment]
@rhorvath37 OK, so there's a few issues here.
First: This issue related to keras import into dl4j (specifically tf.keras), whereas you are using tensorflow import into SameDiff. Separate functionality. So a new issue would have been more appropriate than posting on this one.

Second: The model file you have posted is not a .pb file - it's text format protobuf. SameDiff reads only binary format. You can open it and see for yourself in any text editor.
For more details: https://stackoverflow.com/questions/51948262/difference-between-pb-and-pbtxt-in-tensorflow/51948934

I would suggest saving your model as a frozen binary format protobuf file and trying again.

**Comment by ptamas88** [comment]
Hello @AlexDBlack ,
Thanks for the info about the new snapshot. As I can see in the code, right now there is no way to import embedded models, right? I mean if the layer type is Model.

**Comment by agibsonccc** [comment]
@ptamas88 Sorry for the late response.  For any future readers, please open a new issue if you have a specific model you're trying to import. I remember us getting that working pretty long ago.
